File: android/src/main/kotlin/dev/jerrykhw/flutter_kakao_maps_sdk/view/KakaoMapView.kt
package dev.jerrykhw.flutter_kakao_maps_sdk.view

import android.app.Activity
import android.view.View
import com.kakao.vectormap.Compass
import com.kakao.vectormap.KakaoMap
import com.kakao.vectormap.KakaoMapReadyCallback
import com.kakao.vectormap.LatLng
import com.kakao.vectormap.MapLifeCycleCallback
import com.kakao.vectormap.MapOverlay
import com.kakao.vectormap.MapType
import com.kakao.vectormap.MapView
import com.kakao.vectormap.MapViewInfo
import com.kakao.vectormap.Padding
import com.kakao.vectormap.ScaleBar
import com.kakao.vectormap.camera.CameraPosition
import com.kakao.vectormap.camera.CameraUpdateFactory
import dev.jerrykhw.flutter_kakao_maps_sdk.FlutterKakaoMapsSDKPlugin
import dev.jerrykhw.flutter_kakao_maps_sdk.enum.toMapGravity
import dev.jerrykhw.flutter_kakao_maps_sdk.model.KakaoMapOptions
import dev.jerrykhw.flutter_kakao_maps_sdk.model.toCameraAnimation
import dev.jerrykhw.flutter_kakao_maps_sdk.model.toCompassOptions
import dev.jerrykhw.flutter_kakao_maps_sdk.model.toKakaoMapPosition
import dev.jerrykhw.flutter_kakao_maps_sdk.model.toLatLng
import dev.jerrykhw.flutter_kakao_maps_sdk.model.toPadding
import dev.jerrykhw.flutter_kakao_maps_sdk.model.toPoiOptions
import dev.jerrykhw.flutter_kakao_maps_sdk.model.toScaleBarOptions
import dev.jerrykhw.flutter_kakao_maps_sdk.util.dp
import dev.jerrykhw.flutter_kakao_maps_sdk.util.px
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView
import org.json.JSONObject

internal class KakaoMapView(
    private val activity: Activity,
    private val viewId: Int,
    private var options: KakaoMapOptions,
    private val viewMethodChannel: MethodChannel
) :
    PlatformView {
    private val mapViewContainer: MapView
    private var mapView: KakaoMap? = null
    private var compass: Compass? = null
    private var scaleBar: ScaleBar? = null

    private fun printLog(message: String) {
        FlutterKakaoMapsSDKPlugin.logStreamHandler.sendMessage("KakaoMapView#${viewId}[${options.viewName}] $message")
    }

    private val viewMethodCallHandler =
        MethodChannel.MethodCallHandler { call, result ->
            when (call.method) {
                "dispose" -> dispose(result)
                "moveCamera" -> moveCamera(call.arguments as JSONObject, result)
                "animateCamera" -> animateCamera(call.arguments as JSONObject, result)
                "moveCameraTransform" -> moveCameraTransform(call.arguments as JSONObject, result)
                "animateCameraTransform" -> animateCameraTransform(
                    call.arguments as JSONObject,
                    result
                )

                "setViewInfo" -> setViewInfo(call.arguments as JSONObject, result)
                "showOverlay" -> showOverlay(call.arguments as JSONObject, result)
                "hideOverlay" -> hideOverlay(call.arguments as JSONObject, result)
                "setEnabled" -> setEnabled(call.arguments as JSONObject, result)
                "setBuildingScale" -> setBuildingScale(call.arguments as JSONObject, result)
                "getPadding" -> getPadding(result)
                "setPadding" -> setPadding(call.arguments as JSONObject, result)
                "setLogoPosition" -> setLogoPosition(call.arguments as JSONObject, result)
                "setPoiOptions" -> setPoiOptions(call.arguments as JSONObject, result)
                "setCompassOptions" -> setCompassOptions(call.arguments as JSONObject, result)
                "setScaleBarOptions" -> setScaleBarOptions(call.arguments as JSONObject, result)
                "refresh" -> result.success(null)
                else -> result.notImplemented()
            }
        }

    private fun dispose(result: MethodChannel.Result) {
        printLog("stop")

        mapViewContainer.finish()

        viewMethodChannel.setMethodCallHandler(null)

        result.success(null)
    }

    private fun moveCamera(arguments: JSONObject, result: MethodChannel.Result) {
        printLog("moveCamera")

        val mapView = mapView ?: run {
            result.error("NOT_FOUND_MAPVIEW", "mapview is null", null)
            return
        }

        val cameraPosition = mapView.cameraPosition!!
        var cameraTarget = cameraPosition.position
        var cameraZoomLevel = cameraPosition.zoomLevel
        val cameraHeight = cameraPosition.height
        var cameraRotation = cameraPosition.rotationAngle
        var cameraTilt = cameraPosition.tiltAngle

        if (!arguments.isNull("target")) {
            cameraTarget = arguments.getJSONObject("target").toLatLng()
        }

        if (!arguments.isNull("zoomLevel")) {
            cameraZoomLevel = arguments.getInt("zoomLevel")
        }

        if (!arguments.isNull("rotation")) {
            cameraRotation = arguments.getDouble("rotation")
        }

        if (!arguments.isNull("tilt")) {
            cameraTilt = arguments.getDouble("tilt")
        }

        val cameraUpdate = CameraUpdateFactory.newCameraPosition(
            CameraPosition.from(
                cameraTarget.latitude,
                cameraTarget.longitude,
                cameraZoomLevel,
                cameraTilt,
                cameraRotation,
                cameraHeight,
            ),
        )

        mapView.moveCamera(cameraUpdate)

        result.success(null)
    }

    private fun animateCamera(arguments: JSONObject, result: MethodChannel.Result) {
        printLog("animateCamera")

        val mapView = mapView ?: run {
            result.error("NOT_FOUND_MAPVIEW", "mapview is null", null)
            return
        }

        val cameraPosition = mapView.cameraPosition!!
        var cameraTarget = cameraPosition.position
        var cameraZoomLevel = cameraPosition.zoomLevel
        val cameraHeight = cameraPosition.height
        var cameraRotation = cameraPosition.rotationAngle
        var cameraTilt = cameraPosition.tiltAngle

        if (!arguments.isNull("target")) {
            cameraTarget = arguments.getJSONObject("target").toLatLng()
        }

        if (!arguments.isNull("zoomLevel")) {
            cameraZoomLevel = arguments.getInt("zoomLevel")
        }

        if (!arguments.isNull("rotation")) {
            cameraRotation = arguments.getDouble("rotation")
        }

        if (!arguments.isNull("tilt")) {
            cameraTilt = arguments.getDouble("tilt")
        }

        val cameraAnimationOptions =
            arguments.getJSONObject("cameraAnimationOptions").toCameraAnimation()

        val cameraUpdate = CameraUpdateFactory.newCameraPosition(
            CameraPosition.from(
                cameraTarget.latitude,
                cameraTarget.longitude,
                cameraZoomLevel,
                cameraTilt,
                cameraRotation,
                cameraHeight,
            ),
        )

        mapView.moveCamera(cameraUpdate, cameraAnimationOptions)

        result.success(null)
    }

    private fun moveCameraTransform(arguments: JSONObject, result: MethodChannel.Result) {
        printLog("moveCameraTransform")

        val mapView = mapView ?: run {
            result.error("NOT_FOUND_MAPVIEW", "mapview is null", null)
            return
        }

        val cameraPosition = mapView.cameraPosition!!
        val cameraTarget = cameraPosition.position
        val cameraZoomLevel = cameraPosition.zoomLevel
        val cameraHeight = cameraPosition.height
        val cameraRotation = cameraPosition.rotationAngle
        val cameraTilt = cameraPosition.tiltAngle

        val point = arguments.getJSONObject("point").toLatLng()
        val height = arguments.getDouble("height")
        val rotation = arguments.getDouble("rotation")
        val tilt = arguments.getDouble("tilt")

        val cameraUpdate = CameraUpdateFactory.newCameraPosition(
            CameraPosition.from(
                cameraTarget.latitude + point.latitude,
                cameraTarget.longitude + point.longitude,
                cameraZoomLevel,
                cameraTilt + tilt,
                cameraRotation + rotation,
                cameraHeight + height,
            ),
        )

        mapView.moveCamera(cameraUpdate)

        result.success(null)
    }

    private fun animateCameraTransform(arguments: JSONObject, result: MethodChannel.Result) {
        printLog("animateCameraTransform")

        val mapView = mapView ?: run {
            result.error("NOT_FOUND_MAPVIEW", "mapview is null", null)
            return
        }

            val cameraPosition = mapView.cameraPosition!!
            val cameraTarget = cameraPosition.position
            val cameraZoomLevel = cameraPosition.zoomLevel
            val cameraHeight = cameraPosition.height
            val cameraRotation = cameraPosition.rotationAngle
            val cameraTilt = cameraPosition.tiltAngle

            val point = arguments.getJSONObject("point").toLatLng()
            val height = arguments.getDouble("height")
            val rotation = arguments.getDouble("rotation")
            val tilt = arguments.getDouble("tilt")

            val cameraAnimationOptions =
                arguments.getJSONObject("cameraAnimationOptions").toCameraAnimation()

            val cameraUpdate = CameraUpdateFactory.newCameraPosition(
                CameraPosition.from(
                    cameraTarget.latitude + point.latitude,
                    cameraTarget.longitude + point.longitude,
                    cameraZoomLevel,
                    cameraTilt + tilt,
                    cameraRotation + rotation,
                    cameraHeight + height,
                ),
            )

            mapView.moveCamera(cameraUpdate, cameraAnimationOptions)

        result.success(null)
    }

    private fun setViewInfo(arguments: JSONObject, result: MethodChannel.Result) {
        printLog("setViewInfo")

        val mapView = mapView ?: run {
            result.error("NOT_FOUND_MAPVIEW", "mapview is null", null)
            return
        }

        val appName = arguments.getString("appName")
        val viewInfoName = arguments.getString("viewInfoName")

        mapView.changeMapViewInfo(MapViewInfo.from(appName, viewInfoName))

        result.success(null)
    }

    private fun showOverlay(arguments: JSONObject, result: MethodChannel.Result) {
        printLog("showOverlay")

        val mapView = mapView ?: run {
            result.error("NOT_FOUND_MAPVIEW", "mapview is null", null)
            return
        }

        val overlay = MapOverlay.getEnum(arguments.getString("overlay"))

        mapView.showOverlay(overlay)

        result.success(null)
    }

    private fun hideOverlay(arguments: JSONObject, result: MethodChannel.Result) {
        printLog("hideOverlay")

        val mapView = mapView ?: run {
            result.error("NOT_FOUND_MAPVIEW", "mapview is null", null)
            return
        }

        val overlay = MapOverlay.getEnum(arguments.getString("overlay"))

        mapView.hideOverlay(overlay)

        result.success(null)
    }

    private fun setEnabled(arguments: JSONObject, result: MethodChannel.Result) {
        printLog("setEnabled")

        val mapView = mapView ?: run {
            result.error("NOT_FOUND_MAPVIEW", "mapview is null", null)
            return
        }

        val enabled = arguments.getBoolean("enabled")

        mapView.isVisible = enabled

        result.success(null)
    }

    private fun setBuildingScale(arguments: JSONObject, result: MethodChannel.Result) {
        printLog("setBuildingScale")

        val mapView = mapView ?: run {
            result.error("NOT_FOUND_MAPVIEW", "mapview is null", null)
            return
        }

        val buildingScale = arguments.getDouble("buildingScale").toFloat()

        mapView.buildingHeightScale = buildingScale

        result.success(null)
    }

    private fun getPadding(result: MethodChannel.Result) {
        printLog("getPadding")

        val mapView = mapView ?: run {
            result.error("NOT_FOUND_MAPVIEW", "mapview is null", null)
            return
        }

        val padding = mapView.padding

        val json = JSONObject().apply {
            put("left", padding.left.dp)
            put("top", padding.top.dp)
            put("bottom", padding.bottom.dp)
            put("right", padding.right.dp)
        }

        result.success(json)
    }

    private fun setPadding(arguments: JSONObject, result: MethodChannel.Result) {
        printLog("setPadding")

        val mapView = mapView ?: run {
            result.error("NOT_FOUND_MAPVIEW", "mapview is null", null)
            return
        }

        val padding = arguments.toPadding()

        mapView.setPadding(
            padding.left.px,
            padding.top.px,
            padding.right.px,
            padding.bottom.px,
        )

        result.success(null)
    }

    private fun setLogoPosition(arguments: JSONObject, result: MethodChannel.Result) {
        printLog("setLogoPosition")

        val mapView = mapView ?: run {
            result.error("NOT_FOUND_MAPVIEW", "mapview is null", null)
            return
        }

        val logoPosition = arguments.toKakaoMapPosition()

        mapView.logo?.setPosition(
            logoPosition.alignment.toMapGravity(),
            logoPosition.x.px,
            logoPosition.y.px,
        )

        result.success(null)
    }

    private fun setPoiOptions(arguments: JSONObject, result: MethodChannel.Result) {
        printLog("setPoiOptions")

        val mapView = mapView ?: run {
            result.error("NOT_FOUND_MAPVIEW", "mapview is null", null)
            return
        }

        val poiOptions = arguments.toPoiOptions()

        mapView.setPoiClickable(poiOptions.clickable)
        mapView.setPoiVisible(poiOptions.enabled)
        mapView.setPoiScale(poiOptions.scale)

        result.success(null)
    }

    private fun setCompassOptions(arguments: JSONObject, result: MethodChannel.Result) {
        printLog("setCompassOptions")

        val mapView = mapView ?: run {
            result.error("NOT_FOUND_MAPVIEW", "mapview is null", null)
            return
        }

        val compassOptions = arguments.toCompassOptions()

        compass = mapView.compass
        compass?.let { compass ->
            if (compassOptions.enabled) {
                compass.show()
            } else {
                compass.hide()
            }
            compass.setPosition(
                compassOptions.position.alignment.toMapGravity(),
                compassOptions.position.x.px,
                compassOptions.position.y.px,
            )
        }

        result.success(null)
    }

    private fun setScaleBarOptions(arguments: JSONObject, result: MethodChannel.Result) {
        printLog("setScaleBarOptions")

        val mapView = mapView ?: run {
            result.error("NOT_FOUND_MAPVIEW", "mapview is null", null)
            return
        }

        val scaleBarOptions = arguments.toScaleBarOptions()

        scaleBar = mapView.scaleBar
        scaleBar?.let { scaleBar ->
            if (scaleBarOptions.enabled) {
                scaleBar.show()
            } else {
                scaleBar.hide()
            }
            scaleBar.setPosition(
                scaleBarOptions.position.alignment.toMapGravity(),
                scaleBarOptions.position.x.px,
                scaleBarOptions.position.y.px,
            )
            scaleBar.isAutoHide = scaleBarOptions.autoDisabled
            scaleBar.setFadeInOutTime(
                scaleBarOptions.fadeInOutOptions.fadeInTime,
                scaleBarOptions.fadeInOutOptions.fadeOutTime,
                scaleBarOptions.fadeInOutOptions.retentionTime
            )
        }

        result.success(null)
    }

    init {
        viewMethodChannel.setMethodCallHandler(viewMethodCallHandler)

        mapViewContainer = MapView(activity)

        printLog("start")
        mapViewContainer.start(
            object : MapLifeCycleCallback() {
                override fun onMapResumed() {
                    printLog("onMapResumed")
                }

                override fun onMapPaused() {
                    printLog("onMapPaused")
                }

                override fun onMapDestroy() {
                    printLog("onMapDestroy")
                }

                override fun onMapError(error: Exception) {
                    printLog("onMapError: ${error.message}")
                }
            },
            object : KakaoMapReadyCallback() {
                override fun onMapReady(kakaoMap: KakaoMap) {
                    mapView = kakaoMap

                    mapView?.let { mapView ->
                        // Overlay
                        if (options.overlay != null) {
                            mapView.showOverlay(options.overlay!!)
                        }
                        // Language
                        mapView.setPoiLanguage(options.language)
                        // BuildingScale
                        mapView.buildingHeightScale = options.buildingScale
                        // Padding
                        mapView.setPadding(
                            options.padding.left.px,
                            options.padding.top.px,
                            options.padding.right.px,
                            options.padding.bottom.px,
                        )
                        // LogoPosition
                        mapView.logo?.setPosition(
                            options.logoPosition.alignment.toMapGravity(),
                            options.logoPosition.x.px,
                            options.logoPosition.y.px,
                        )
                        // PoiOptions
                        mapView.setPoiClickable(options.poiOptions.clickable)
                        mapView.setPoiVisible(options.poiOptions.enabled)
                        mapView.setPoiScale(options.poiOptions.scale)
                        // CompassOptions
                        compass = mapView.compass
                        compass?.let { compass ->
                            if (options.compassOptions.enabled) {
                                compass.show()
                            } else {
                                compass.hide()
                            }
                            compass.setPosition(
                                options.compassOptions.position.alignment.toMapGravity(),
                                options.compassOptions.position.x.px,
                                options.compassOptions.position.y.px,
                            )
                        }
                        // ScaleBarOptions
                        scaleBar = mapView.scaleBar
                        scaleBar?.let { scaleBar ->
                            if (options.scaleBarOptions.enabled) {
                                scaleBar.show()
                            } else {
                                scaleBar.hide()
                            }
                            scaleBar.setPosition(
                                options.scaleBarOptions.position.alignment.toMapGravity(),
                                options.scaleBarOptions.position.x.px,
                                options.scaleBarOptions.position.y.px,
                            )
                            scaleBar.isAutoHide = options.scaleBarOptions.autoDisabled
                            scaleBar.setFadeInOutTime(
                                options.scaleBarOptions.fadeInOutOptions.fadeInTime,
                                options.scaleBarOptions.fadeInOutOptions.fadeOutTime,
                                options.scaleBarOptions.fadeInOutOptions.retentionTime
                            )
                        }

                        viewMethodChannel.invokeMethod("onMapReady", null)
                    }
                }

                override fun getViewName(): String {
                    return options.viewName
                }

                override fun getMapViewInfo(): MapViewInfo {
                    return MapViewInfo.from(options.appName, MapType.getEnum(options.viewInfoName))
                }

                override fun getPosition(): LatLng {
                    return options.defaultPosition
                }

                override fun getZoomLevel(): Int {
                    return options.defaultLevel
                }

                override fun isVisible(): Boolean {
                    return options.enabled
                }
            },
        )
    }

    override fun getView(): View {
        return mapViewContainer
    }

    override fun dispose() = Unit
}

File: ios/Classes/view/KakaoMapView.swift
import Flutter
import KakaoMapsSDK

class KakaoMapView: NSObject, FlutterPlatformView, MapControllerDelegate {
    private let mapViewContainer: KMViewContainer
    private let mapController: KMController
    private var mapView: KakaoMap?
    
    private var options: KakaoMapOptions
    private let viewMethodChannel: FlutterMethodChannel
    
    private let viewId: Int64
    
    private var auth: Bool
    
    private func printLog(_ message: String) {
        FlutterKakaoMapsSDKPlugin.logStreamHandler.sendMessage("KakaoMapView#\(viewId)[\(options.viewName)] \(message)")
    }
    
    private func viewMethodCallHandler(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "dispose": dispose(onSuccess: result)
        case "moveCamera": moveCamera(arguments: call.arguments as! NSDictionary, onSuccess: result)
        case "animateCamera": animateCamera(arguments: call.arguments as! NSDictionary, onSuccess: result)
        case "moveCameraTransform": moveCameraTransform(arguments: call.arguments as! NSDictionary, onSuccess: result)
        case "animateCameraTransform": animateCameraTransform(arguments: call.arguments as! NSDictionary, onSuccess: result)
        case "setViewInfo": setViewInfo(arguments: call.arguments as! NSDictionary, onSuccess: result)
        case "showOverlay": showOverlay(arguments: call.arguments as! NSDictionary, onSuccess: result)
        case "hideOverlay": hideOverlay(arguments: call.arguments as! NSDictionary, onSuccess: result)
        case "setEnabled": setEnabled(arguments: call.arguments as! NSDictionary, onSuccess: result)
        case "setBuildingScale": setBuildingScale(arguments: call.arguments as! NSDictionary, onSuccess: result)
        case "getPadding": getPadding(onSuccess: result)
        case "setPadding": setPadding(arguments: call.arguments as! NSDictionary, onSuccess: result)
        case "setLogoPosition": setLogoPosition(arguments: call.arguments as! NSDictionary, onSuccess: result)
        case "setPoiOptions": setPoiOptions(arguments: call.arguments as! NSDictionary, onSuccess: result)
        case "setCompassOptions": setCompassOptions(arguments: call.arguments as! NSDictionary, onSuccess: result)
        case "setScaleBarOptions": setScaleBarOptions(arguments: call.arguments as! NSDictionary, onSuccess: result)
        case "refresh": refresh(onSuccess: result)
        default: result(FlutterMethodNotImplemented)
        }
    }
    
    func dispose(onSuccess: @escaping (Any?) -> ()) {
        printLog("stopEngine")
        
        mapController.stopRendering()
        mapController.stopEngine()
        
        viewMethodChannel.setMethodCallHandler(nil)
        
        onSuccess(nil)
    }
    
    func moveCamera(arguments: NSDictionary, onSuccess: @escaping (Any?) -> ()) {
        printLog("moveCamera")
        
        if let mapView = self.mapView {
            let cameraPosition = CameraUpdate.make(mapView: mapView).cameraPosition!
            var cameraTarget = cameraPosition.targetPoint
            var cameraZoomLevel = mapView.zoomLevel
            var cameraRotation = cameraPosition.rotation
            var cameraTilt = cameraPosition.tilt
            
            let target = arguments["target"] as? NSDictionary
            if let target = target {
                cameraTarget = target.toMapPoint()
            }
            
            let zoomLevel = arguments["zoomLevel"] as? Int
            if let zoomLevel = zoomLevel {
                cameraZoomLevel = zoomLevel
            }
            
            let rotation = arguments["rotation"] as? Double
            if let rotation = rotation {
                cameraRotation = rotation
            }
            
            let tilt = arguments["tilt"] as? Double
            if let tilt = tilt {
                cameraTilt = tilt
            }
            
            let cameraUpdate = CameraUpdate.make(target: cameraTarget, zoomLevel: cameraZoomLevel, rotation: cameraRotation, tilt: cameraTilt, mapView: mapView)
            
            mapView.moveCamera(cameraUpdate)
        }
        
        onSuccess(nil)
    }
    
    func animateCamera(arguments: NSDictionary, onSuccess: @escaping (Any?) -> ()) {
        printLog("animateCamera")
        
        if let mapView = self.mapView {
            let cameraPosition = CameraUpdate.make(mapView: mapView).cameraPosition!
            var cameraTarget = cameraPosition.targetPoint
            var cameraZoomLevel = mapView.zoomLevel
            var cameraRotation = cameraPosition.rotation
            var cameraTilt = cameraPosition.tilt
            
            let target = arguments["target"] as? NSDictionary
            if let target = target {
                cameraTarget = target.toMapPoint()
            }
            
            let zoomLevel = arguments["zoomLevel"] as? Int
            if let zoomLevel = zoomLevel {
                cameraZoomLevel = zoomLevel
            }
            
            let rotation = arguments["rotation"] as? Double
            if let rotation = rotation {
                cameraRotation = rotation
            }
            
            let tilt = arguments["tilt"] as? Double
            if let tilt = tilt {
                cameraTilt = tilt
            }
            
            let cameraAnimationOptions = (arguments["cameraAnimationOptions"] as! NSDictionary).toCameraAnimationOptions()
            
            let cameraUpdate = CameraUpdate.make(target: cameraTarget, zoomLevel: cameraZoomLevel, rotation: cameraRotation, tilt: cameraTilt, mapView: mapView)
            
            mapView.animateCamera(cameraUpdate: cameraUpdate, options: cameraAnimationOptions)
        }
        
        onSuccess(nil)
    }
    
    func moveCameraTransform(arguments: NSDictionary, onSuccess: @escaping (Any?) -> ()) {
        printLog("moveCameraTransform")
        
        if let mapView = self.mapView {
            let point = (arguments["point"] as! NSDictionary).toCameraTransformDelta()
            let height = arguments["height"] as! Double
            let rotation = arguments["rotation"] as! Double
            let tilt = arguments["tilt"] as! Double
            
            let cameraUpdate = CameraUpdate.make(transform: CameraTransform(deltaPos: point, deltaHeight: height, deltaRotation: rotation, deltaTilt: tilt))
            
            mapView.moveCamera(cameraUpdate)
        }
        
        onSuccess(nil)
    }
    
    func animateCameraTransform(arguments: NSDictionary, onSuccess: @escaping (Any?) -> ()) {
        printLog("animateCameraTransform")
        
        if let mapView = self.mapView {
            let point = (arguments["point"] as! NSDictionary).toCameraTransformDelta()
            let height = arguments["height"] as! Double
            let rotation = arguments["rotation"] as! Double
            let tilt = arguments["tilt"] as! Double
            let cameraAnimationOptions = (arguments["cameraAnimationOptions"] as! NSDictionary).toCameraAnimationOptions()
            
            let cameraUpdate = CameraUpdate.make(transform: CameraTransform(deltaPos: point, deltaHeight: height, deltaRotation: rotation, deltaTilt: tilt))
            
            mapView.animateCamera(cameraUpdate: cameraUpdate, options: cameraAnimationOptions)
        }
        
        onSuccess(nil)
    }
    
    func setViewInfo(arguments: NSDictionary, onSuccess: @escaping (Any?) -> ()) {
        printLog("setViewInfo")
        
        let appName = arguments["appName"] as! String
        let viewInfoName = arguments["viewInfoName"] as! String
        options.appName = appName
        options.viewInfoName = viewInfoName
        
        mapView?.changeViewInfo(appName: appName, viewInfoName: viewInfoName)
        
        onSuccess(nil)
    }
    
    func showOverlay(arguments: NSDictionary, onSuccess: @escaping (Any?) -> ()) {
        printLog("showOverlay")
        
        let overlay = arguments["overlay"] as! String
        options.overlay = overlay
        
        mapView?.showOverlay(overlay)
        
        onSuccess(nil)
    }
    
    func hideOverlay(arguments: NSDictionary, onSuccess: @escaping (Any?) -> ()) {
        printLog("hideOverlay")
        
        let overlay = arguments["overlay"] as! String
        
        mapView?.hideOverlay(overlay)
        
        onSuccess(nil)
    }
    
    func setEnabled(arguments: NSDictionary, onSuccess: @escaping (Any?) -> ()) {
        printLog("setEnabled")
        
        let enabled = arguments["enabled"] as! Bool
        options.enabled = enabled
        
        mapView?.isEnabled = enabled
        
        onSuccess(nil)
    }
    
    func setBuildingScale(arguments: NSDictionary, onSuccess: @escaping (Any?) -> ()) {
        printLog("setBuildingScale")
        
        let buildingScale = Float(arguments["buildingScale"] as! Double)
        options.buildingScale = buildingScale
        
        mapView?.buildingScale = buildingScale
        
        onSuccess(nil)
    }
    
    func getPadding(onSuccess: @escaping (Any?) -> ()) {
        printLog("getPadding")
        
        let margins = mapView?.margins ?? UIEdgeInsets()
        
        onSuccess(["left": margins.left, "top": margins.top, "right": margins.right, "bottom":margins.bottom])
    }
    
    func setPadding(arguments: NSDictionary, onSuccess: @escaping (Any?) -> ()) {
        printLog("setPadding")
        
        let padding = arguments.toUIEdgeInsets()
        options.padding = padding
        
        mapView?.setMargins(padding)
        
        onSuccess(nil)
    }
    
    func setLogoPosition(arguments: NSDictionary, onSuccess: @escaping (Any?) -> ()) {
        printLog("setLogoPosition")
        
        let logoPosition = arguments.toKakaoMapPosition()
        options.logoPosition = logoPosition
        
        mapView?.setLogoPosition(origin: logoPosition.alignment.toGuiAlignment(), position: CGPoint(x:logoPosition.x, y: logoPosition.y))
        
        onSuccess(nil)
    }
    
    func setPoiOptions(arguments: NSDictionary, onSuccess: @escaping (Any?) -> ()) {
        printLog("setPoiOptions")
        
        let poiOptions = arguments.toPoiOptions()
        options.poiOptions = poiOptions
        
        mapView?.poiClickable = poiOptions.clickable
        mapView?.setPoiEnabled(poiOptions.enabled)
        mapView?.poiScale = poiOptions.scale
        
        onSuccess(nil)
    }
    
    func setCompassOptions(arguments: NSDictionary, onSuccess: @escaping (Any?) -> ()) {
        printLog("setCompassOptions")
        
        let compassOptions = arguments.toCompassOptions()
        options.compassOptions = compassOptions
        
        if(compassOptions.enabled) {
            mapView?.showCompass()
        } else {
            mapView?.hideCompass()
        }
        mapView?.setCompassPosition(origin: compassOptions.position.alignment.toGuiAlignment(), position: CGPoint(x: compassOptions.position.x, y: compassOptions.position.y))
        
        onSuccess(nil)
    }
    
    func setScaleBarOptions(arguments: NSDictionary, onSuccess: @escaping (Any?) -> ()) {
        printLog("setScaleBarOptions")
        
        let scaleBarOptions = arguments.toScaleBarOptions()
        options.scaleBarOptions = scaleBarOptions
        
        if(scaleBarOptions.enabled) {
            mapView?.showScaleBar()
        } else {
            mapView?.hideScaleBar()
        }
        mapView?.setScaleBarPosition(origin: scaleBarOptions.position.alignment.toGuiAlignment(), position: CGPoint(x:scaleBarOptions.position.x, y: scaleBarOptions.position.y))
        mapView?.setScaleBarAutoDisappear(scaleBarOptions.autoDisabled)
        mapView?.setScaleBarFadeInOutOption(scaleBarOptions.fadeInOutOptions)
        
        onSuccess(nil)
    }
    
    func refresh(onSuccess: @escaping (Any?) -> ()) {
        printLog("refresh")
        
        mapView?.refresh()
        
        onSuccess(nil)
    }
    
    init(
        frame: CGRect,
        viewIdentifier viewId: Int64,
        options: KakaoMapOptions,
        viewMethodChannel: FlutterMethodChannel
    ) {
        self.auth = false
        self.viewId = viewId
        self.options = options
        self.viewMethodChannel = viewMethodChannel
        
        mapViewContainer = KMViewContainer(frame: frame)
        mapController = KMController(viewContainer: mapViewContainer)!
        
        if mapViewContainer.proMotionDisplay == true {
            mapController.proMotionSupport = true
        }
        
        super.init()
        
        viewMethodChannel.setMethodCallHandler(viewMethodCallHandler)
        
        mapController.delegate = self
        
        printLog("init")
    }
    
    func authenticationSucceeded() {
        printLog("auth success")
        
        if auth == false {
            auth = true
            self.mapController.startEngine()
            self.mapController.startRendering()
        }
    }
    
    func authenticationFailed(_ errorCode: Int, desc: String) {
        printLog("auth error code: \(errorCode)")
        printLog("auth desc: \(desc)")
        
        auth = false
        
        switch errorCode {
        case 499:
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
                self.printLog("auth retry")
                
                self.mapController.authenticate()
            }
            break;
        default:
            break;
        }
    }
    
    func addViews() {
        let mapViewInfo = MapviewInfo(viewName: options.viewName, appName: options.appName, viewInfoName: options.viewInfoName, defaultPosition: options.defaultPosition, defaultLevel: options.defaultLevel, enabled: options.enabled)
        
        if mapController.addView(mapViewInfo) == Result.OK {
            mapView = mapController.getView(options.viewName) as? KakaoMap
            
            // Overlay
            if(options.overlay != nil) {
                mapView?.showOverlay(options.overlay!)
            }
            // Language
            mapView?.setLanguage(options.language)
            // BuildingScale
            mapView?.buildingScale = options.buildingScale
            // Padding
            mapView?.setMargins(options.padding)
            // LogoPosition
            mapView?.setLogoPosition(origin: options.logoPosition.alignment.toGuiAlignment(), position: CGPoint(x: options.logoPosition.x, y: options.logoPosition.y))
            // PoiOptions
            mapView?.poiClickable = options.poiOptions.clickable
            mapView?.setPoiEnabled(options.poiOptions.enabled)
            mapView?.poiScale = options.poiOptions.scale
            // CompassOptions
            if(options.compassOptions.enabled) {
                mapView?.showCompass()
            } else {
                mapView?.hideCompass()
            }
            mapView?.setCompassPosition(origin: options.compassOptions.position.alignment.toGuiAlignment(), position: CGPoint(x: options.compassOptions.position.x, y: options.compassOptions.position.y))
            // ScaleBarOptions
            if(options.scaleBarOptions.enabled) {
                mapView?.showScaleBar()
            } else {
                mapView?.hideScaleBar()
            }
            mapView?.setScaleBarPosition(origin: options.scaleBarOptions.position.alignment.toGuiAlignment(), position: CGPoint(x: options.scaleBarOptions.position.x, y: options.scaleBarOptions.position.y))
            mapView?.setScaleBarAutoDisappear(options.scaleBarOptions.autoDisabled)
            mapView?.setScaleBarFadeInOutOption(options.scaleBarOptions.fadeInOutOptions)
            
        }
    }
    
    func view() -> UIView {
        printLog("initEngine")
        mapController.initEngine()
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
            if self.auth {
                self.printLog("startEngine")
                self.mapController.startEngine()
                self.mapController.startRendering()
            } else {
                self.printLog("authenticate")
                self.mapController.authenticate()
            }
        }
        
        return mapViewContainer
    }
}
